modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java

```diff
@@ -0,0 +1,2728 @@
+                    userNewArrayFunctionRefNode.getLocation(), targetType.getTargetType(), "this", name, 0);
+            valueType = targetType.getTargetType();
+            semanticScope.putDecoration(userNewArrayFunctionRefNode, new ReferenceDecoration(ref));
+        }
+
+        semanticScope.putDecoration(userNewArrayFunctionRefNode, new ValueType(valueType));
+    }
+
+    @Override
+    public void visitSymbol(ESymbol userSymbolNode, SemanticScope semanticScope) {
+        boolean read = semanticScope.getCondition(userSymbolNode, Read.class);
+        boolean write = semanticScope.getCondition(userSymbolNode, Write.class);
+        String symbol = userSymbolNode.getSymbol();
+        Class<?> staticType = semanticScope.getScriptScope().getPainlessLookup().canonicalTypeNameToType(symbol);
+
+        if (staticType != null)  {
+            if (write) {
+                throw userSymbolNode.createError(new IllegalArgumentException("invalid assignment: " +
+                        "cannot write a value to a static type [" + PainlessLookupUtility.typeToCanonicalTypeName(staticType) + "]"));
+            }
+
+            if (read == false) {
+                throw userSymbolNode.createError(new IllegalArgumentException("not a statement: " +
+                        "static type [" + PainlessLookupUtility.typeToCanonicalTypeName(staticType) + "] not used"));
+            }
+
+            semanticScope.putDecoration(userSymbolNode, new StaticType(staticType));
+        } else if (semanticScope.isVariableDefined(symbol)) {
+            if (read == false && write == false) {
+                throw userSymbolNode.createError(new IllegalArgumentException("not a statement: variable [" + symbol + "] not used"));
+            }
+
+            Location location = userSymbolNode.getLocation();
+            Variable variable = semanticScope.getVariable(location, symbol);
+
+            if (write && variable.isFinal()) {
+                throw userSymbolNode.createError(new IllegalArgumentException("Variable [" + variable.getName() + "] is read-only."));
+            }
+
+            Class<?> valueType = variable.getType();
+            semanticScope.putDecoration(userSymbolNode, new ValueType(valueType));
+        } else {
+            semanticScope.putDecoration(userSymbolNode, new PartialCanonicalTypeName(symbol));
+        }
+    }
+
+    @Override
+    public void visitDot(EDot userDotNode, SemanticScope semanticScope) {
+        boolean read = semanticScope.getCondition(userDotNode, Read.class);
+        boolean write = semanticScope.getCondition(userDotNode, Write.class);
+
+        if (read == false && write == false) {
+            throw userDotNode.createError(new IllegalArgumentException("not a statement: result of dot operator [.] not used"));
+        }
+
+        ScriptScope scriptScope = semanticScope.getScriptScope();
+        String index = userDotNode.getIndex();
```

Comment:
Consider moving this definition closer to it's use.

Reply:
Moved these as far down as possible.